src/Mvc/Mvc.RazorPages/src/Infrastructure/PageActionInvokerProvider.cs

```diff
@@ -139,7 +138,19 @@ public void OnProvidersExecuting(ActionInvokerProviderContext context)
             IFilterMetadata[] filters;
             if (!cache.Entries.TryGetValue(actionDescriptor, out var cacheEntry))
             {
-                actionContext.ActionDescriptor = _loader.Load(actionDescriptor);
+                CompiledPageActionDescriptor compiledPageActionDescriptor;
+                var endpointFeature = actionContext.HttpContext.Features.Get<IEndpointFeature>();
+                if (endpointFeature != null)
+                {
+                    // With endpoint routing, PageLoaderMatcherPolicy should have already produced a CompiledPageActionDescriptor.
+                    compiledPageActionDescriptor = (CompiledPageActionDescriptor)actionDescriptor;
+                }
+                else
+                {
+                    compiledPageActionDescriptor = _loader.LoadAsync(actionDescriptor).GetAwaiter().GetResult();
```

Comment:
Fixing this is a bunch more work. We could leave it less nice without endpoint routing

Reply:
I think it's it's fine. We can't really do anything about the fact that the old code path is sync.